### DISCUSSION TITLE
Status Update: 2022-12-21

### DISCUSSION THREAD

**Discussion by haixuanTao** [dora-rs/dora#161] [Status Update]
:christmas_tree:

**Comment by haixuanTao** [comment]
## dora-drives

- Adding native installation of dora-drives as an alternative to docker. I have updated the whole documentation to add native launch of dora-drives.
- Adding Carla Leaderboard and scenario runner after Bob gave me the different version: https://github.com/dora-rs/dora-drives/pull/23 . 

- Connected dora-rs to Carla leaderboard-2.0
    - Data from the leaderboard to the dataflow and back is ok.
    - Issues with Leaderboard Map: https://github.com/carla-simulator/leaderboard/issues/144 -> Issue is bypassed by direct use of Carla API
  
[![IMAGE ALT TEXT](http://img.youtube.com/vi/sCF2CIHIkb0/0.jpg)](http://www.youtube.com/watch?v=sCF2CIHIkb0 "Video Title")

> Tested on: GeForce GTX 1050 Ti Mobile
> Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz

Currently connected operator:
- [x] Leaderboard agent to send input data and receive control to manipulate car.
- [x] Yolov5
- [x] Obstacle Location using Lidar
- [x] Carla GPS operator able to retrieve GPS waypoints to go to Goal location
- [x] Frenet Optimal Planner to avoid obstacle
- [x] PID to derive acceleration, steering and brake from waypoints and position 
- [x] Visualisation
- [ ] yolop and car behaviour with lane
- [ ] Strong_sort to track and predict obstacle movement.


Software version chosen:

|Software|Version Tested|Installation Guide|
|--------|-------|------------------|
|Linux|Ubuntu 20.04.5 LTS||
|Miniconda|22.11.1|Installation in doc|
|Pytorch|1.11|Installation in doc|
|Carla|Carla Leaderboard|Installation in `scripts/install.sh`. Version: [Leaderboard Version](https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/Leaderboard/CARLA_Leaderboard_20.tar.gz)|
|NVIDIA Driver|515.86.01||
|CUDA|11.7||
|dora-rs|0.1.1|Installation in doc|

### Work investigation and next step.

- Docker containerisation. Small issue with Vulkan driver with the new Carla version
- Traffic Light management in Pylot is very handtuned. Might not be a good idea to import as it might not work in real case scenario. Good obstacle avoidance might be preferable See: https://github.com/erdos-project/pylot/blob/cbc937c1cd36e2086165061b7707b52c85f30f8b/pylot/planning/world.py#L329-L443
- Stop sign management seems to be still in progress. See: https://github.com/erdos-project/pylot/blob/cbc937c1cd36e2086165061b7707b52c85f30f8b/pylot/map/hd_map.py#L163-L177 
- Next step is probably about **lane management** and **obstacle tracking and prediction**.

**Reply by haixuanTao** [reply]
-> Make sure Docker run and get simulation score. Docker container
-> Test the Online platform.
  -> Account. AWS ICR. Dedicated Repo. 
  -> Docker run container. Container should contain everything. 
  -> Cronjob capture run.

Map, Town, and scenarios? One track work in process.

-> How do they capture the simulation score format?

-> Communication channel ?

**Comment by haixuanTao** [comment]
## dora

Bumped minor version dora-rs v0.1.2 to fix CPU usage issue and have unique Linux version. https://github.com/dora-rs/dora/pull/160 with:

- Fixed infinite loop in the coordinator. https://github.com/dora-rs/dora/pull/155
- simplified release process. https://github.com/dora-rs/dora/pull/157
- Generic linux distribution. https://github.com/dora-rs/dora/pull/159

**Comment by phil-opp** [comment]
## Progress on `dora-daemon`

- Connect daemon to dora-coordinator and implement command to spawn nodes
- Map node outputs to node inputs of downstream nodes
- Use reference-counted drop tokens to free shared memory regions once all receivers are done with them
- Implement support for timer interval inputs in daemon
- Update coordinator to forward `dora stop` commands to connected daemons
- Notify coordinator when a dataflow is finished -> ensure that `dora list` is up-to-date
- Update Python node API for new daemon design

### Next Steps

- Update Python dataflow example for new node API
- Reduce input/output latency by reusing shared memory regions instead of reallocating them
- Update remaning `dora` CLI command to work with new daemon-based design (e.g., `dora up`, `dora destroy`)

**Reply by haixuanTao** [reply]
daemon <-> coordinator, grpc? Protobuf definition?
